### ISSUE TITLE
Referencing name not defined yet allowed by mypy, yet error at runtime.

### ISSUE THREAD

**Issue by MarioIshac** [python/mypy#17610]
**To Reproduce**

```python
from collections.abc import Callable

class A:
    @staticmethod
    def b(c: Callable[[], None]) -> Callable[[], None]:
        return c
    
    @A.b
    @staticmethod
    def c() -> None:
        pass
```

[mypy playground](https://mypy-play.net/?mypy=latest&python=3.12&flags=strict&gist=f5d9c559c428ac03dc556d77e2bf0c53)

**Expected Behavior**

Invoking `mypy` catches error.

**Actual Behavior**

```
Success: no issues found in 1 source file
```

However, at runtime:

```
NameError: name 'A' is not defined
```
